### PR TITLE
Add Causal Mask support for Softmax

### PR DESCRIPTION
**Pull request by yugaoTT**
add causal mask for `scale_mask_softmax_in_place` and `scale_mask_softmax`, supports both sharded and interleaved

to use causal mask instead of mask, add one extra arg `is_causal_mask=True` when calling the softmax function

### CODE REVIEW

File: tests/tt_eager/python_api_testing/unit_testing/test_softmax.py [hunk 1]
```diff
@@ -19,151 +19,97 @@
 from models.utility_functions import torch2tt_tensor, tt2torch_tensor, pad_by_zero
 
 
-@skip_for_wormhole_b0()
-@pytest.mark.parametrize("inplace", [True, False])
-def test_softmax(device, inplace):
-    torch.manual_seed(0)
-    sm_op = ttl.operations.primary.softmax_in_place if inplace else ttl.tensor.softmax
+# @skip_for_wormhole_b0()
```
Comment (arakhmati) [inline tests/tt_eager/python_api_testing/unit_testing/test_softmax.py:22]: Did you mean to comment out the tests?

If no longer need this code, we should just delete it

File: tt_eager/tt_dnn/op_library/softmax/softmax_op.hpp [hunk 2]
```diff
@@ -47,6 +47,7 @@ struct Softmax {
     const bool inplace;
     const MemoryConfig output_mem_config;
     const tt::operations::primary::transformers::SoftmaxProgramConfig program_config;
+    const bool is_causal_mask;
```
Comment (arakhmati) [inline tt_eager/tt_dnn/op_library/softmax/softmax_op.hpp:50]: Do we need this attribute?

Can it be inferred by the op instead based on the mask shape?

**Review by arakhmati — APPROVED** [review]

File: tt_eager/tt_dnn/op_library/softmax/softmax_op.hpp (hunk 2 above)
Comment (arakhmati) [inline tt_eager/tt_dnn/op_library/softmax/softmax_op.hpp:50]: Do we need this attribute?

Can it be inferred by the op instead based on the mask shape?

**Review by arakhmati — CHANGES_REQUESTED** [review]

**Review by mywoodstock — APPROVED** [review]

File: tests/tt_eager/python_api_testing/unit_testing/test_softmax_sharded.py [hunk 3]
```diff
@@ -54,19 +59,37 @@ def test_softmax(device, in_dtype, cb_dtype, in0_mem_config):
...
+        attention_mask = attention_mask.reshape(batch, 1, -1, 32)
+        # attention_mask32 = tilize_to_list(pad_weight(attention_mask))
+        # attention_mask_t = ttl.tensor.Tensor(
+        #     attention_mask32,
+        #     [batch, 1, 32, 384],
+        #     ttl.tensor.DataType.BFLOAT16,
+        #     ttl.tensor.Layout.TILE,
+        #     device,
+        #     ttl.tensor.MemoryConfig(ttl.tensor.TensorMemoryLayout.INTERLEAVED, ttl.tensor.BufferType.L1),
+        # )
```
Comment (tt-aho) [inline tests/tt_eager/python_api_testing/unit_testing/test_softmax_sharded.py:76]: Remove if this isn't used

**Review by arakhmati — APPROVED** [review]
